src/app/services/canvas.service.ts

```diff
@@ -1549,4 +1577,78 @@ export class CanvasService {
             data: { assignment: { published: published } }
         });
     }
+
+    public async getAssignmentGradingType(
+        courseId: string,
+        assignmentId: string
+    ): Promise<keyof typeof CanvasGradingType> {
+        const data = await this.apiRequest(`/api/v1/courses/${courseId}/assignments/${assignmentId}`);
+        if (typeof data['grading_type'] != 'string') {
+            throw new Error('Invalid data');
+        }
+        return data['grading_type'] as keyof typeof CanvasGradingType;
+    }
+
+    public async getAssignmentGradingTypeAndPointsPossible(
```

Comment:
We could (and should) always make attributes that does not guaranteed to exist as optional attributes (via `t.union([t.null, t. undefined, <actual type>])`). I'm suggesting this since every REST API call will return these attributes regardless whether you need it or not, but things might change with GraphQL.

Reply:
Agreed. Maybe something similar to the mixin pattern would help, but it would be a pain to make the type system work as expected when the combination of multiple extensions are possible.

Reply:
TypeScript provides a utility type `Partial<T>`, which returns a type by making all properties of `T` optional. The opposite is `Required<T>`. Combining these two utility types with type intersection `&` could achieve what you want. The hard part is that if we have multiple flags (e.g., `hasGradingType` and `hasPublishState`), changing the function return value in correspondance of these flags could be difficult (but I think it's still feasible with TypeScript).